### PR TITLE
Discover new fields when detecting changes in update case

### PR DESCRIPTION
**Pull request by msftcoderdjw**
Fix issues: #450

### CODE REVIEW

**Review by LeiQL — APPROVED** [review]

**Review by RemindD — APPROVED** [review]

**Review by iwangjintian — APPROVED** [review]

**Comment by msftcoderdjw** [comment]
Will discuss this PR in weekly eng sync.

File: api/pkg/apis/v1alpha1/model/validationrule.go [hunk 1]
```diff
@@ -82,6 +82,9 @@ func detectChanges(properties []PropertyDesc, oldName string, newName string, ol
 			regexpPattern := strings.ReplaceAll(escapedPattern, `\*`, ".*")
 			// Compile the regular expression
 			regexpObject := regexp.MustCompile("^" + regexpPattern + "$")
+			if len(oldValues) != len(newValues) {
```
Comment (Haishi2016) [inline api/pkg/apis/v1alpha1/model/validationrule.go:85]: I think this is over simplified. let's say the pattern is "a*"  and old value is ["ab"] while the new value is ["ab", "cd"]. Although the number of properties differ, "cd" is not matched by the pattern hence it should not be considered.

Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/model/validationrule.go:85]: Yes, this is not the final version. I am testing the behavior of skipIfMissing. After confirming the behavior, I will fix this PR and convert it to public.

Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/model/validationrule.go]: Fixed the logic. The compareProperties will work like this,

```go
func compareProperties(c PropertyDesc, old map[string]interface{}, new map[string]interface{}, key string) bool {
	v, ook := old[key]
	nv, nok := new[key]
	if c.PropChanged != nil {
		return c.PropChanged(v, nv)
	}

	if ook && nok {
		// case 1: key exists in both old and new
		// compare the values, if different, return true
		if !compareStrings(fmt.Sprintf("%v", v), fmt.Sprintf("%v", nv), c.IgnoreCase, c.PrefixMatch) {
			return true
		} else {
			return false
		}
	} else if !ook && !nok {
		// case 2: key does not exist in both old and new
		// return false to indicate no change
		return false
	} else {
		// case 3: one of them is missing
		// if the property is optional, no matter it doesn't exist in old or new, return false to indicate no change
		if !c.SkipIfMissing {
			return true
		} else {
			return false
		}
	}
}
```

File: api/pkg/apis/v1alpha1/model/validationrule.go [hunk 2]
```diff
@@ -164,20 +181,28 @@ func compareProperties(c PropertyDesc, old map[string]interface{}, new map[strin
...
+
+	if ook && nok {
+		// case 1: key exists in both old and new
+		// compare the values, if different, return true
+		if !compareStrings(fmt.Sprintf("%v", v), fmt.Sprintf("%v", nv), c.IgnoreCase, c.PrefixMatch) {
 			return true
+		} else {
+			return false
 		}
+	} else if !ook && !nok {
```
[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/model/validationrule.go:193]: If key never existed in both old and new maps, we will treat return false to indicate no change.

[the remark this answers is not in the document]
Comment (Haishi2016) [reply api/pkg/apis/v1alpha1/model/validationrule.go:193]: I meant will this condition ever be hit?

[the remark this answers is not in the document]
Comment (msftcoderdjw) [reply api/pkg/apis/v1alpha1/model/validationrule.go:193]: Yes, if provider defines ComponentValidationRule.ChangeDetectionProperties and ComponentValidationRule.ChangeDetectionMetadata to include field name, but this field doesn't show up in old and new maps. 

https://github.com/eclipse-symphony/symphony/blob/dd6285bd0264983ec8fa967c54f467e0d264e924/api/pkg/apis/v1alpha1/model/validationrule.go#L93C1-L101C5

**Review by Haishi2016 — APPROVED** [review]

**Pull request merged**